File: server/src/utils/spotifyAuthUtils.js
import crypto from 'crypto';

// Function to generate a code verifier
export function generateCodeVerifier() {
    return crypto.randomBytes(32).toString('hex');
}

// Function to generate a code challenge
export function generateCodeChallenge(codeVerifier) {
    const hash = crypto.createHash('sha256');
    hash.update(codeVerifier);
    const codeChallenge = hash
        .digest('base64')
        .replace(/\+/g, '-')
        .replace(/\//g, '_')
        .replace(/=+$/, '');
    return codeChallenge;
}
//     } else {

File: backend/spotify_auth/spotify_auth_utils.py
# spotify_auth_utils.py

import base64
import os
import hashlib

def generate_code_verifier() -> str:
    """Generate a random code verifier."""
    random_bytes = os.urandom(64)
    code_verifier = base64.urlsafe_b64encode(random_bytes).decode('utf-8').rstrip('=')
    return code_verifier

def generate_code_challenge(code_verifier: str) -> str:
    """Generate a code challenge from the code verifier."""
    code_verifier_bytes = code_verifier.encode('utf-8')
    sha256_hash = hashlib.sha256(code_verifier_bytes).digest()
    code_challenge = base64.urlsafe_b64encode(sha256_hash).decode('utf-8').rstrip('=')
    return code_challenge
